File: backend/src/routes/internships.ts
import express from 'express';
import { PrismaClient } from '@prisma/client';
import { authenticateToken, AuthRequest } from '../middleware/auth';
import { optionalAuth } from '../middleware/auth';
import { 
  validateInternshipCreate, 
  handleValidationErrors 
} from '../utils/validation';

const router = express.Router();
const prisma = new PrismaClient();

// Get all internships with optional filtering
router.get('/', optionalAuth, async (req, res) => {
  try {
    const { 
      page = 1, 
      limit = 10, 
      location, 
      remote, 
      company, 
      skills,
      interests,
      search 
    } = req.query;

    const skip = (Number(page) - 1) * Number(limit);
    const take = Number(limit);

    // Build where clause
    const where: any = {
      active: true
    };

    if (location) {
      where.location = {
        contains: location as string,
        mode: 'insensitive'
      };
    }

    if (remote !== undefined) {
      where.remote = remote === 'true';
    }

    if (company) {
      where.company = {
        name: {
          contains: company as string,
          mode: 'insensitive'
        }
      };
    }

    if (search) {
      where.OR = [
        {
          title: {
            contains: search as string,
            mode: 'insensitive'
          }
        },
        {
          description: {
            contains: search as string,
            mode: 'insensitive'
          }
        }
      ];
    }

    if (skills) {
      const skillArray = (skills as string).split(',');
      where.skills = {
        some: {
          skill: {
            name: {
              in: skillArray,
              mode: 'insensitive'
            }
          }
        }
      };
    }

    if (interests) {
      const interestArray = (interests as string).split(',');
      where.interests = {
        some: {
          interest: {
            name: {
              in: interestArray,
              mode: 'insensitive'
            }
          }
        }
      };
    }

    const internships = await prisma.internship.findMany({
      where,
      skip,
      take,
      include: {
        company: {
          select: {
            id: true,
            name: true,
            logo: true,
            location: true,
            industry: true
          }
        },
        skills: {
          include: {
            skill: true
          }
        },
        interests: {
          include: {
            interest: true
          }
        },
        _count: {
          select: {
            applications: true,
            savedBy: true
          }
        }
      },
      orderBy: {
        postedDate: 'desc'
      }
    });

    const total = await prisma.internship.count({ where });

    res.json({
      internships,
      pagination: {
        page: Number(page),
        limit: Number(limit),
        total,
        pages: Math.ceil(total / Number(limit))
      }
    });

  } catch (error) {
    console.error('Get internships error:', error);
    res.status(500).json({
      error: 'Internal server error'
    });
  }
});

// Get internship by ID
router.get('/:id', optionalAuth, async (req, res) => {
  try {
    const { id } = req.params;

    const internship = await prisma.internship.findUnique({
      where: { id },
      include: {
        company: true,
        skills: {
          include: {
            skill: true
          }
        },
        interests: {
          include: {
            interest: true
          }
        },
        _count: {
          select: {
            applications: true,
            savedBy: true
          }
        }
      }
    });

    if (!internship) {
      return res.status(404).json({
        error: 'Internship not found'
      });
    }

    // Check if user has saved this internship
    let isSaved = false;
    if (req.user) {
      const savedInternship = await prisma.savedInternship.findUnique({
        where: {
          userId_internshipId: {
            userId: req.user.id,
            internshipId: id
          }
        }
      });
      isSaved = !!savedInternship;
    }

    res.json({
      internship: {
        ...internship,
        isSaved
      }
    });

  } catch (error) {
    console.error('Get internship error:', error);
    res.status(500).json({
      error: 'Internal server error'
    });
  }
});

// Create new internship (admin only)
router.post('/', authenticateToken, validateInternshipCreate, handleValidationErrors, async (req: AuthRequest, res) => {
  try {
    const {
      title,
      description,
      companyId,
      location,
      duration,
      salary,
      requirements,
      teamSize,
      deadline,
      remote = false,
      skillIds = [],
      interestIds = []
    } = req.body;

    // Check if company exists
    const company = await prisma.company.findUnique({
      where: { id: companyId }
    });

    if (!company) {
      return res.status(404).json({
        error: 'Company not found'
      });
    }

    // Create internship
    const internship = await prisma.internship.create({
      data: {
        title,
        description,
        companyId,
        location,
        duration,
        salary,
        requirements,
        teamSize,
        deadline: deadline ? new Date(deadline) : null,
        remote,
        skills: {
          create: skillIds.map((skillId: string) => ({
            skillId,
            required: true
          }))
        },
        interests: {
          create: interestIds.map((interestId: string) => ({
            interestId
          }))
        }
      },
      include: {
        company: true,
        skills: {
          include: {
            skill: true
          }
        },
        interests: {
          include: {
            interest: true
          }
        }
      }
    });

    res.status(201).json({
      message: 'Internship created successfully',
      internship
    });

  } catch (error) {
    console.error('Create internship error:', error);
    res.status(500).json({
      error: 'Internal server error'
    });
  }
});

// Update internship
router.put('/:id', authenticateToken, async (req: AuthRequest, res) => {
  try {
    const { id } = req.params;
    const updateData = req.body;

    // Check if internship exists
    const existingInternship = await prisma.internship.findUnique({
      where: { id }
    });

    if (!existingInternship) {
      return res.status(404).json({
        error: 'Internship not found'
      });
    }

    // Update internship
    const updatedInternship = await prisma.internship.update({
      where: { id },
      data: {
        ...updateData,
        deadline: updateData.deadline ? new Date(updateData.deadline) : undefined
      },
      include: {
        company: true,
        skills: {
          include: {
            skill: true
          }
        },
        interests: {
          include: {
            interest: true
          }
        }
      }
    });

    res.json({
      message: 'Internship updated successfully',
      internship: updatedInternship
    });

  } catch (error) {
    console.error('Update internship error:', error);
    res.status(500).json({
      error: 'Internal server error'
    });
  }
});

// Delete internship
router.delete('/:id', authenticateToken, async (req: AuthRequest, res) => {
  try {
    const { id } = req.params;

    // Check if internship exists
    const existingInternship = await prisma.internship.findUnique({
      where: { id }
    });

    if (!existingInternship) {
      return res.status(404).json({
        error: 'Internship not found'
      });
    }

    // Soft delete by setting active to false
    await prisma.internship.update({
      where: { id },
      data: { active: false }
    });

    res.json({
      message: 'Internship deleted successfully'
    });

  } catch (error) {
    console.error('Delete internship error:', error);
    res.status(500).json({
      error: 'Internal server error'
    });
  }
});

// Save/unsave internship
router.post('/:id/save', authenticateToken, async (req: AuthRequest, res) => {
  try {
    const { id } = req.params;
    const userId = req.user!.id;

    // Check if internship exists
    const internship = await prisma.internship.findUnique({
      where: { id }
    });

    if (!internship) {
      return res.status(404).json({
        error: 'Internship not found'
      });
    }

    // Check if already saved
    const existingSaved = await prisma.savedInternship.findUnique({
      where: {
        userId_internshipId: {
          userId,
          internshipId: id
        }
      }
    });

    if (existingSaved) {
      // Remove from saved
      await prisma.savedInternship.delete({
        where: {
          userId_internshipId: {
            userId,
            internshipId: id
          }
        }
      });

      res.json({
        message: 'Internship removed from saved',
        saved: false
      });
    } else {
      // Add to saved
      await prisma.savedInternship.create({
        data: {
          userId,
          internshipId: id
        }
      });

      res.json({
        message: 'Internship saved successfully',
        saved: true
      });
    }

  } catch (error) {
    console.error('Save internship error:', error);
    res.status(500).json({
      error: 'Internal server error'
    });
  }
});

// Get saved internships for user
router.get('/saved/list', authenticateToken, async (req: AuthRequest, res) => {
  try {
    const userId = req.user!.id;

    const savedInternships = await prisma.savedInternship.findMany({
      where: { userId },
      include: {
        internship: {
          include: {
            company: {
              select: {
                id: true,
                name: true,
                logo: true,
                location: true
              }
            },
            skills: {
              include: {
                skill: true
              }
            },
            interests: {
              include: {
                interest: true
              }
            },
            _count: {
              select: {
                applications: true
              }
            }
          }
        }
      },
      orderBy: {
        savedAt: 'desc'
      }
    });

    res.json({
      savedInternships: savedInternships.map(item => ({
        ...item.internship,
        savedAt: item.savedAt
      }))
    });

  } catch (error) {
    console.error('Get saved internships error:', error);
    res.status(500).json({
      error: 'Internal server error'
    });
  }
});

export default router;

File: backend/routes/internships.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity, verify_jwt_in_request
from extensions import db
from models import Internship, Company, InternshipSkill, InternshipInterest, SavedInternship
import uuid
from datetime import datetime

internships_bp = Blueprint('internships', __name__)

def get_current_user_id():
    """Get current user ID if authenticated, otherwise return None"""
    try:
        verify_jwt_in_request()
        return get_jwt_identity()
    except:
        return None

@internships_bp.route('/', methods=['GET'])
def get_internships():
    try:
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))
        location = request.args.get('location')
        remote = request.args.get('remote')
        company = request.args.get('company')
        skills = request.args.get('skills')
        interests = request.args.get('interests')
        search = request.args.get('search')
        
        skip = (page - 1) * limit
        
        # Build query
        query = Internship.query.filter_by(active=True)
        
        if location:
            query = query.filter(Internship.location.ilike(f'%{location}%'))
        
        if remote is not None:
            query = query.filter(Internship.remote==(remote.lower() == 'true'))
        
        if company:
            query = query.join(Company).filter(Company.name.ilike(f'%{company}%'))
        
        if search:
            query = query.filter(
                db.or_(
                    Internship.title.ilike(f'%{search}%'),
                    Internship.description.ilike(f'%{search}%')
                )
            )
        
        if skills:
            skill_list = skills.split(',')
            query = query.join(InternshipSkill).join(Skill).filter(
                Skill.name.in_(skill_list)
            )
        
        if interests:
            interest_list = interests.split(',')
            query = query.join(InternshipInterest).join(Interest).filter(
                Interest.name.in_(interest_list)
            )
        
        # Get total count
        total = query.count()
        
        # Get paginated results
        internships = query.options(
            db.joinedload(Internship.company),
            db.joinedload(Internship.skills).joinedload(InternshipSkill.skill),
            db.joinedload(Internship.interests).joinedload(InternshipInterest.interest)
        ).offset(skip).limit(limit).all()
        
        # Get current user ID for saved status
        user_id = get_current_user_id()
        
        # Format results
        internship_list = []
        for internship in internships:
            internship_dict = internship.to_dict()
            internship_dict['company'] = internship.company.to_dict()
            internship_dict['skills'] = [is_obj.to_dict() for is_obj in internship.skills]
            internship_dict['interests'] = [ii_obj.to_dict() for ii_obj in internship.interests]
            
            # Check if saved by current user
            if user_id:
                is_saved = SavedInternship.query.filter_by(
                    user_id=user_id,
                    internship_id=internship.id
                ).first()
                internship_dict['is_saved'] = bool(is_saved)
            else:
                internship_dict['is_saved'] = False
            
            internship_list.append(internship_dict)
        
        return jsonify({
            'internships': internship_list,
            'pagination': {
                'page': page,
                'limit': limit,
                'total': total,
                'pages': (total + limit - 1) // limit
            }
        }), 200
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

@internships_bp.route('/<internship_id>', methods=['GET'])
def get_internship(internship_id):
    try:
        internship = Internship.query.options(
            db.joinedload(Internship.company),
            db.joinedload(Internship.skills).joinedload(InternshipSkill.skill),
            db.joinedload(Internship.interests).joinedload(InternshipInterest.interest)
        ).get(internship_id)
        
        if not internship:
            return jsonify({'error': 'Internship not found'}), 404
        
        internship_dict = internship.to_dict()
        internship_dict['company'] = internship.company.to_dict()
        internship_dict['skills'] = [is_obj.to_dict() for is_obj in internship.skills]
        internship_dict['interests'] = [ii_obj.to_dict() for ii_obj in internship.interests]
        
        # Check if saved by current user
        user_id = get_current_user_id()
        if user_id:
            is_saved = SavedInternship.query.filter_by(
                user_id=user_id,
                internship_id=internship_id
            ).first()
            internship_dict['is_saved'] = bool(is_saved)
        else:
            internship_dict['is_saved'] = False
        
        return jsonify({'internship': internship_dict}), 200
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

@internships_bp.route('/', methods=['POST'])
@jwt_required()
def create_internship():
    try:
        data = request.get_json()
        
        # Validate required fields
        required_fields = ['title', 'description', 'company_id', 'location', 'duration']
        for field in required_fields:
            if not data.get(field):
                return jsonify({'error': f'{field} is required'}), 400
        
        # Check if company exists
        company = Company.query.get(data['company_id'])
        if not company:
            return jsonify({'error': 'Company not found'}), 404
        
        # Create internship
        internship = Internship(
            id=str(uuid.uuid4()),
            title=data['title'],
            description=data['description'],
            company_id=data['company_id'],
            location=data['location'],
            duration=data['duration'],
            salary=data.get('salary'),
            requirements=data.get('requirements', []),
            team_size=data.get('team_size'),
            deadline=datetime.fromisoformat(data['deadline']) if data.get('deadline') else None,
            remote=data.get('remote', False)
        )
        
        db.session.add(internship)
        db.session.flush()  # Get the ID
        
        # Add skills
        skill_ids = data.get('skill_ids', [])
        for skill_id in skill_ids:
            internship_skill = InternshipSkill(
                id=str(uuid.uuid4()),
                internship_id=internship.id,
                skill_id=skill_id,
                required=True
            )
            db.session.add(internship_skill)
        
        # Add interests
        interest_ids = data.get('interest_ids', [])
        for interest_id in interest_ids:
            internship_interest = InternshipInterest(
                id=str(uuid.uuid4()),
                internship_id=internship.id,
                interest_id=interest_id
            )
            db.session.add(internship_interest)
        
        db.session.commit()
        
        # Return created internship
        internship_dict = internship.to_dict()
        internship_dict['company'] = company.to_dict()
        internship_dict['skills'] = [is_obj.to_dict() for is_obj in internship.skills]
        internship_dict['interests'] = [ii_obj.to_dict() for ii_obj in internship.interests]
        
        return jsonify({
            'message': 'Internship created successfully',
            'internship': internship_dict
        }), 201
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Internal server error'}), 500

@internships_bp.route('/<internship_id>', methods=['PUT'])
@jwt_required()
def update_internship(internship_id):
    try:
        internship = Internship.query.get(internship_id)
        
        if not internship:
            return jsonify({'error': 'Internship not found'}), 404
        
        data = request.get_json()
        
        # Update allowed fields
        allowed_fields = ['title', 'description', 'location', 'duration', 'salary', 
                         'requirements', 'team_size', 'remote', 'active']
        
        for field in allowed_fields:
            if field in data:
                setattr(internship, field, data[field])
        
        if 'deadline' in data and data['deadline']:
            internship.deadline = datetime.fromisoformat(data['deadline'])
        
        db.session.commit()
        
        # Return updated internship
        internship_dict = internship.to_dict()
        internship_dict['company'] = internship.company.to_dict()
        internship_dict['skills'] = [is_obj.to_dict() for is_obj in internship.skills]
        internship_dict['interests'] = [ii_obj.to_dict() for ii_obj in internship.interests]
        
        return jsonify({
            'message': 'Internship updated successfully',
            'internship': internship_dict
        }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Internal server error'}), 500

@internships_bp.route('/<internship_id>', methods=['DELETE'])
@jwt_required()
def delete_internship(internship_id):
    try:
        internship = Internship.query.get(internship_id)
        
        if not internship:
            return jsonify({'error': 'Internship not found'}), 404
        
        # Soft delete by setting active to false
        internship.active = False
        db.session.commit()
        
        return jsonify({'message': 'Internship deleted successfully'}), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Internal server error'}), 500

@internships_bp.route('/<internship_id>/save', methods=['POST'])
@jwt_required()
def save_internship(internship_id):
    try:
        user_id = get_jwt_identity()
        
        # Check if internship exists
        internship = Internship.query.get(internship_id)
        if not internship:
            return jsonify({'error': 'Internship not found'}), 404
        
        # Check if already saved
        existing_saved = SavedInternship.query.filter_by(
            user_id=user_id,
            internship_id=internship_id
        ).first()
        
        if existing_saved:
            # Remove from saved
            db.session.delete(existing_saved)
            db.session.commit()
            
            return jsonify({
                'message': 'Internship removed from saved',
                'saved': False
            }), 200
        else:
            # Add to saved
            saved_internship = SavedInternship(
                id=str(uuid.uuid4()),
                user_id=user_id,
                internship_id=internship_id
            )
            
            db.session.add(saved_internship)
            db.session.commit()
            
            return jsonify({
                'message': 'Internship saved successfully',
                'saved': True
            }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Internal server error'}), 500

@internships_bp.route('/saved/list', methods=['GET'])
@jwt_required()
def get_saved_internships():
    try:
        user_id = get_jwt_identity()
        
        saved_internships = SavedInternship.query.filter_by(user_id=user_id).options(
            db.joinedload(SavedInternship.internship).joinedload(Internship.company),
            db.joinedload(SavedInternship.internship).joinedload(Internship.skills).joinedload(InternshipSkill.skill),
            db.joinedload(SavedInternship.internship).joinedload(Internship.interests).joinedload(InternshipInterest.interest)
        ).order_by(SavedInternship.saved_at.desc()).all()
        
        saved_list = []
        for saved in saved_internships:
            internship = saved.internship
            internship_dict = internship.to_dict()
            internship_dict['company'] = internship.company.to_dict()
            internship_dict['skills'] = [is_obj.to_dict() for is_obj in internship.skills]
            internship_dict['interests'] = [ii_obj.to_dict() for ii_obj in internship.interests]
            internship_dict['saved_at'] = saved.saved_at.isoformat()
            
            saved_list.append(internship_dict)
        
        return jsonify({'saved_internships': saved_list}), 200
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

@internships_bp.route('/companies', methods=['GET'])
def get_companies():
    try:
        companies = Company.query.order_by(Company.name.asc()).all()
        
        return jsonify({
            'companies': [company.to_dict() for company in companies]
        }), 200
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500
